Apps/Src/Tasks.c

```diff
@@ -40,28 +41,95 @@ CPU_STK DebugDump_Stk[TASK_DEBUG_DUMP_STACK_SIZE];
+
+// Variables to store error codes, stored and cleared in task error assert functions
+error_code_t Error_ReadTritium = T_NONE;  // Initialized to no error
+error_code_t Error_ReadCarCAN = /*READCARCAN_ERR_NONE*/ 0; // TODO: change this back to the error 
+error_code_t Error_UpdateDisplay = UPDATEDISPLAY_ERR_NONE;
+
+extern const PinInfo_t PINS_LOOKARR[]; // For GPIO writes. Externed from Minions Driver C file.
 
 /**
- * Global Variables
- */
-fault_bitmap_t FaultBitmap = FAULT_NONE;
-os_error_loc_t OSErrLocBitmap = OS_NONE_LOC;
+ * Error assertion-related functions
+*/
 
-void _assertOSError(uint16_t OS_err_loc, OS_ERR err)
+void _assertOSError(os_error_loc_t OS_err_loc, OS_ERR err)
 {
     if (err != OS_ERR_NONE)
     {
-        FaultBitmap |= FAULT_OS;
-        OSErrLocBitmap |= OS_err_loc;
+        arrayMotorKill(); // Turn off contactors and turn on the brakelight to indicate an emergency
+        Display_Error(OS_err_loc, err); // Display the location and error code
+        while(1){;} //nonrecoverable
 
-        OSSemPost(&FaultState_Sem4, OS_OPT_POST_1, &err);
-        EnterFaultState();
     }
 }
 
+/**
+ * @brief Assert a task error by locking the scheduler (if necessary), displaying a fault screen,
+ * and jumping to the error's specified callback function. 
+ * Called by task-specific error-assertion functions that are also responsible for setting the error variable.
+ * @param errorLoc the task from which the error originated. Note: should be taken out when last task pointer is integrated
+ * @param faultCode the enum for the specific error that happened
+ * @param errorCallback a callback function to a handler for that specific error, 
+ * @param schedLock whether or not to lock the scheduler to ensure the error is handled immediately
+ * @param nonrecoverable whether or not to kill the motor, display the fault screen, and enter an infinite while loop
+*/
+void assertTaskError(os_error_loc_t errorLoc, error_code_t errorCode, callback_t errorCallback, error_scheduler_lock_opt_t lockSched, error_recov_opt_t nonrecoverable) {
+    OS_ERR err;
+
+    if (lockSched == OPT_LOCK_SCHED) { // We want this error to be handled immediately without other tasks being able to interrupt
+        OSSchedLock(&err);
+        assertOSError(OS_TASKS_LOC, err);
+    }
+
+    // Set the location error variable
+    OSErrLocBitmap = errorLoc; 
+
+    if (nonrecoverable == OPT_NONRECOV) {
+        arrayMotorKill(); // Apart from while loop because killing the motor is more important
+        Display_Error(errorLoc, errorCode); // Needs to happen before callback so that tasks can change the screen
+        // (ex: readCarCAN and evac screen for BPS trip)
+    }
+
+
+    if (errorCallback != NULL) {
+        errorCallback(); // Run a handler for this error that was specified in another task file
+    }
+    
+
+    if (nonrecoverable == OPT_NONRECOV) { // enter an infinite while loop
+        while(1);
+    }
```

Comment:
Maybe we can print something here so we can get feedback in the test rather then only using GDB to see if we are stuck here

Reply:
Yes, maybe print all of the faults we have for each application.